bson-kotlinx/src/main/kotlin/org/bson/codecs/kotlinx/BsonDecoder.kt

```diff
@@ -0,0 +1,260 @@
+import kotlinx.serialization.descriptors.PolymorphicKind
+import kotlinx.serialization.descriptors.PrimitiveKind
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.SerialKind
+import kotlinx.serialization.descriptors.StructureKind
+import kotlinx.serialization.descriptors.elementDescriptors
+import kotlinx.serialization.encoding.AbstractDecoder
+import kotlinx.serialization.encoding.CompositeDecoder
+import kotlinx.serialization.encoding.CompositeDecoder.Companion.DECODE_DONE
+import kotlinx.serialization.encoding.CompositeDecoder.Companion.UNKNOWN_NAME
+import kotlinx.serialization.modules.SerializersModule
+import org.bson.AbstractBsonReader
+import org.bson.BsonInvalidOperationException
+import org.bson.BsonReader
+import org.bson.BsonType
+import org.bson.BsonValue
+import org.bson.codecs.BsonValueCodec
+import org.bson.codecs.DecoderContext
+import org.bson.types.ObjectId
+
+/**
+ * The BsonDecoder interface
+ *
+ * For custom serialization handlers
+ */
+public sealed interface BsonDecoder {
+
+    /** @return the decoded ObjectId */
+    public fun decodeObjectId(): ObjectId
+    /** @return the decoded BsonValue */
+    public fun decodeBsonValue(): BsonValue
+
+    /** @return the BsonReader */
+    public fun reader(): BsonReader
+}
+
+@ExperimentalSerializationApi
+internal open class DefaultBsonDecoder(
+    internal val reader: AbstractBsonReader,
+    override val serializersModule: SerializersModule,
+    internal val configuration: BsonConfiguration
+) : BsonDecoder, AbstractDecoder() {
+
+    companion object {
+        val validKeyKinds = setOf(PrimitiveKind.STRING, PrimitiveKind.CHAR, SerialKind.ENUM)
+        val bsonValueCodec = BsonValueCodec()
+    }
+
+    private var elementsIsNullableIndexes: BooleanArray? = null
+
+    private fun initElementNullsIndexes(descriptor: SerialDescriptor) {
+        if (elementsIsNullableIndexes != null) return
+        val elementIndexes = BooleanArray(descriptor.elementsCount)
+        descriptor.elementDescriptors.withIndex().forEach {
+            elementIndexes[it.index] = !descriptor.isElementOptional(it.index) && it.value.isNullable
+        }
+        elementsIsNullableIndexes = elementIndexes
+    }
+
+    @Suppress("ReturnCount")
+    override fun decodeElementIndex(descriptor: SerialDescriptor): Int {
+        initElementNullsIndexes(descriptor)
+
+        val name: String? =
+            when (reader.state!!) {
+                AbstractBsonReader.State.NAME -> reader.readName()
+                AbstractBsonReader.State.VALUE -> reader.currentName
+                AbstractBsonReader.State.TYPE -> {
+                    reader.readBsonType()
+                    return decodeElementIndex(descriptor)
+                }
+                AbstractBsonReader.State.END_OF_DOCUMENT,
+                AbstractBsonReader.State.END_OF_ARRAY -> {
+                    val indexOfNullableElement = elementsIsNullableIndexes!!.indexOfFirst { it }
```

Comment:
It's better to write:
val isNullableIndexes = elementsIsNullableIndexes ?: error("elementsIsNullableIndexes may not be null)
val indexOfNullableElement = isNullableIndexes.indexOfFirst { it }

And also remove !! of line 98